File: matlab/ch10/gaopt.m
function [xopt,fopt,iter] = gaopt(fun,xl,xu,nb,ps,ng,mp)
% gaopt.m: maximization by the genetic algorithm
% Inputs:
% fun: objective functions
% xl,xu: lower and upper bounds on x
% nb: number of binary digits
% ps: population size
% ng: number of generations
% mp: mutation probability
% Outputs:
% xopt: optimal point
% fopt: objective function value at x=xopt
% iter: number of function evaluations
% Example:
% xl=-6*[1 1 1 1]; xu=6*[1 1 1 1]; nb=8; ps=40; ng=50; mp=0.02;
% f = @(x) (x(1)+10*x(2))^2+5*(x(3)-x(4))^2+(x(2)-2*x(3))^4+10*(x(1)-x(4))^4;
% [xopt,fopt,iter] = gaopt(f,xl,xu,nb,ps,ng,mp)
n = length(xu);
Pn = round(rand(ps,n*nb)); % initial population
for kg = 1:ng
ft = fitness(fun,ps,nb,n,xl,xu,Pn); [fmax,kmax] = max(ft);
if (kg == 1)
fbest = fmax;
for k = 1:n
ab = num2str(Pn(kmax,nb*(k-1)+1:nb*k));
adec = bin2dec(ab); % convert binary string to decimal integer
xbest(k) = xl(k) + (xu(k)-xl(k))/(2^nb - 1)*adec;
end
else
if (fmax > fbest)
fbest = fmax;
for k = 1:n
ab = num2str(Pn(kmax,nb*(k-1)+1:nb*k)); adec = bin2dec(ab);
xbest(k) = xl(k) + (xu(k)-xl(k))/(2^nb - 1)*adec;
end
end
end
[frk,ft] = frank(ps, ft); % scaling
for k = 1: ps % shuffling: roulette wheel
ik = frk(k); ptemp = ps; ktemp = k; am(ik) = 2*(ptemp + 1 - ktemp)/ (ptemp*(ptemp + 1));
end
for k = 2: ps, ptemp = am(k - 1); am(k) = ptemp + am(k); end
%
for k = 1:ps % selection
kstr1 = roul(ps,am); kstr2 = roul(ps,am); kchd = cros(Pn,nb,n,kstr1, kstr2);
kchd = mutat(kchd,mp,nb,n);
for j = 1: nb*n, Pn(k, j) = kchd(j); end
end
end
xopt = xbest; fopt = fbest; iter = ng*ps;
end
function ft = fitness(fun,ps,nb,n,xl,xu,Pn) % function evaluation
for k = 1:ps
for j = 1:n
a = num2str(Pn(k,nb*(j-1) + 1:nb*j)); adec = bin2dec(a);
x(j) = xl(j) + (xu(j)-xl(j))/(2^nb - 1)*adec;
end
f = fun(x); ft(k) = f;
end
end

function [frk,ft] = frank(ps, ft)
for k = 1:ps, frk(k) = k; end
for k = 1:ps-1
temp = ft(k); ktemp = k;
for j = k+1:ps
if (ft(j) > temp), temp = ft(j); ktemp = j; end
end
jtemp = frk(ktemp); frk(ktemp) = frk(k); frk(k) = jtemp;
ftemp = ft(ktemp); ft(ktemp) = ft(k); ft(k) = ftemp;
end
end
function kstr = roul(ps,am) % shuffling operation
temp = rand;
for k = 1:ps
while (1)
if (temp > am(k)), break; end
kstr = k; return;
end
end
end
function kchd = cros(Pn,nb,n,kstr1, kstr2) % crossover operation
for j = 1:n
jp = nb*(j - 1); nc = floor(rand*nb + 0.5);
if (nc == 0)
for k = 1:nb, kchd(k+jp) = Pn(kstr2, k+jp); end
elseif (nc == nb)
for k = 1:nb, kchd(k+jp) = Pn(kstr1, k+jp); end
else
for k = 1: nc, kchd(k+jp) = Pn(kstr1, k+jp); end
for k = nc+1:nb, kchd(k+jp) = Pn(kstr2, k+jp); end
end
end
end
function [kchd] = mutat(kchd,mp,nb,n) % mutation operation
for k = 1:nb*n
if (rand <= mp)
if (kchd(k) == 1), kchd(k) = 0;
else, kchd(k) = 1;
end
end
end
end

File: python/ch10/gaopt.py
import numpy as np

def gaopt(fun, xl, xu, nb, ps, ng, mp):
    """
    gaopt: maximization by the genetic algorithm
    """
    n = len(xu)
    # 초기 인구 생성 (0과 1의 랜덤 행렬)
    Pn = np.random.randint(2, size=(ps, n * nb))
    
    xbest = np.zeros(n)
    fbest = -np.inf
    
    for kg in range(ng):
        ft = fitness(fun, ps, nb, n, xl, xu, Pn)
        fmax_idx = np.argmax(ft)
        fmax = ft[fmax_idx]
        
        # 최적값 업데이트
        if kg == 0 or fmax > fbest:
            fbest = fmax
            # 최적해 디코딩
            for k in range(n):
                bits = Pn[fmax_idx, nb*k : nb*(k+1)]
                adec = int("".join(map(str, bits)), 2)
                xbest[k] = xl[k] + (xu[k] - xl[k]) / (2**nb - 1) * adec
        
        # 스케일링 및 룰렛 휠 계산
        frk, ft_sorted = frank(ps, ft)
        am = np.zeros(ps)
        for k in range(ps):
            ik = int(frk[k]) - 1 # 0-indexed
            am[ik] = 2 * (ps + 1 - (k + 1)) / (ps * (ps + 1))
        
        # 누적 확률 계산
        am = np.cumsum(am)
        
        # 선택 및 교차/변이
        new_Pn = np.zeros_like(Pn)
        for k in range(ps):
            kstr1 = roul(ps, am)
            kstr2 = roul(ps, am)
            kchd = cros(Pn, nb, n, kstr1, kstr2)
            kchd = mutat(kchd, mp, nb, n)
            new_Pn[k, :] = kchd
        Pn = new_Pn
        
    return xbest, fbest, ng * ps

def fitness(fun, ps, nb, n, xl, xu, Pn):
    ft = np.zeros(ps)
    for k in range(ps):
        x = np.zeros(n)
        for j in range(n):
            bits = Pn[k, nb*j : nb*(j+1)]
            adec = int("".join(map(str, bits)), 2)
            x[j] = xl[j] + (xu[j] - xl[j]) / (2**nb - 1) * adec
        ft[k] = fun(x)
    return ft

def frank(ps, ft):
    indices = np.arange(1, ps + 1)
    # ft 기준으로 내림차순 정렬
    sorted_idx = np.argsort(ft)[::-1]
    ft_sorted = ft[sorted_idx]
    frk = indices[sorted_idx]
    return frk, ft_sorted

def roul(ps, am):
    temp = np.random.rand()
    for k in range(ps):
        if temp <= am[k]:
            return k
    return ps - 1

def cros(Pn, nb, n, kstr1, kstr2):
    kchd = np.zeros(nb * n, dtype=int)
    for j in range(n):
        jp = nb * j
        nc = np.random.randint(0, nb + 1)
        if nc == 0:
            kchd[jp:jp+nb] = Pn[kstr2, jp:jp+nb]
        elif nc == nb:
            kchd[jp:jp+nb] = Pn[kstr1, jp:jp+nb]
        else:
            kchd[jp:jp+nc] = Pn[kstr1, jp:jp+nc]
            kchd[jp+nc:jp+nb] = Pn[kstr2, jp+nc:jp+nb]
    return kchd

def mutat(kchd, mp, nb, n):
    mask = np.random.rand(nb * n) <= mp
    kchd[mask] = 1 - kchd[mask]
    return kchd